Add operators for +,-,*,/ for `half` with other scalar types

nilsfriess opened the pull request:
This also required removing the `explicit` keyword for the constructors and adding three new constructors for `unsigned int`, `unsigned long` and `long`.

nilsfriess committed:
Add operators for +,-,*,/ for `half` with other scalar types

This also required removing the `explicit` keyword for the
constructors and adding three new constructors for `unsigned int`,
`unsigned long` and `long`.

illuhad commented:
Can you comment more on the use case? What is the code you attempt to make work?

nilsfriess commented:
Things like
```
sycl::half a(1);
float b = 1;
auto res = a * b;
```
currently do not compile because the multiplication is ambiguous (the compiler could first create a `half` from `b` and then use `operator*(const half& a, const half& b)` or convert `a` to `float` and use the built-in `operator*(float, float)`).
There are a few places in the CTS where scalars are multiplied with/divided by `half` and to compile the CTS it would probably suffice to just implement the operators for `float`. I just added the other ones for completeness.

nilsfriess committed:
Add operators for `long long` and `unsigned long long`

nilsfriess committed:
Remove all constructors and conversion operators expect for `float`

We can rely on implicit conversion here, because according to
https://en.cppreference.com/w/cpp/language/static_cast, the static
cast performs the same conversion as an implicit conversion (and
implicit conversions from {integral types, double} -> float are valid
according to https://en.cppreference.com/w/cpp/language/implicit_conversion)

nilsfriess committed:
Manually create half in `operators(half, half)`

Since we removed the half(fp16::half_storage) constructors, we must
explicitly construct the half that is returned. Unfortunately, the
macro `__hipsycl_backend_switch` contains an if statement for the host
pass in the generic backend, so we cannot just write
`auto data = __hipsycl_backend_switch(...)`.

nilsfriess committed:
Fix half tests for new operators

nilsfriess committed:
Simplify operators and fix return type (half operators always return half)

nilsfriess committed:
Fix type of result buffer in tests

illuhad commented:
The SYCL WG agreed that `half` should indeed only have the `float` conversion operator: https://github.com/KhronosGroup/SYCL-CTS/issues/559#issuecomment-1481490154

illuhad commented on include/hipSYCL/sycl/libkernel/half.hpp:
```diff
@@ -64,56 +68,58 @@ class half {
-      return fp16::builtin_add(a._data, b._data),
-      return fp16::builtin_add(a._data, b._data))
+      data = fp16::builtin_add(a._data, b._data),
+      data = fp16::builtin_add(a._data, b._data));
```
Couldn't you simplify this along the lines of:
```c++
return detail::create_half(__hipsycl_backend_switch(
  fp16::builtin_add(a._data, b._data),
  ....
));
```

nilsfriess replied:
Unfortunately not, because `__hipsycl_backend_switch` adds semicolons to its arguments, so this becomes something like 
```
return detail::create_half(fp16::builtin_add(a._data, b._data););
```
which does not compile (this could of course be fixed by removing the semicolons in the definition of `__hipsycl_backend_switch` and adding them in the places where the macro is used). However, for the sscp backend, `__hipsycl_backend_switch` becomes an if statement, which again doesn't compile, and we can't just change that to a ternary operator because sometimes `__hipsycl_backend_switch` is used with return statements as its arguments and these are not allowed as expressions in the ternary operator.

illuhad replied:
Ugh... even if SSCP might in principle work with ternary operator, it probably won't work with nvc++, where we need `if target()` branching.. So you are totally right :(

illuhad commented:
Can you resolve merge conflicts?
